fixes #17198, DFA failure on large case stmts

saem opened the pull request:
This alters the DFA control flow graph generation for case statments.
Gotos are now generated as a chained link, this ensures that evaluation
of variant branches collapses as early as possible, without hitting the
2k call limit.

Additionally, while testing this issue a further problem arose wherein
recent removal of nil seq support broke the ast/len proc. Fixed along
with a test.

saem commented:
@Clyybber if you're about this is the approach if done via DFA CFG generation.

Araq commented on compiler/ast.nim:
```diff
@@ -1128,7 +1128,17 @@ proc discardSons*(father: PNode)
 
 proc len*(n: Indexable): int {.inline.} =
-  result = n.sons.len
+  when n is PNode:
```
Instead use `safeLen` where required, don't slow down good old `len` please.

saem replied:
I can do that, but note that all checkSons, loops, etc... are now prone to break as they no longer guard on nil like they used to.

I know it's used _everywhere_, something better would be to somehow force the nil check as I believe the variant's branch will still be nil. 🤔 Don't suppose you got a sneak trick to do that? 🥷

Araq replied:
I don't understand. All I'm asking for is that you don't touch `ast.nim`.

saem replied:
If one provides a malformed AST now it will likely crash the compiler, without providing an ill formed AST message.

I can double check the symptom but should be replicable with the test I provided on a version before this change and after the removal of non-nil sequences.

Araq replied:
Malformed ASTs are not `nil`, they are `nkEmpty` (and hopefully soon `nkError`).

saem replied:
I know that and it's too early to even detect it, IIRC we blindly sem the return of a macro, so all of that is not at all going to help. :(

The issue is that semIf presently crashes as the malformed AST is passed in. All the usual guards, including checkMinSons will also crash. This will be pervasive throughout the compiler and not detected by tests because we don't fuzz on macro input.

No matter what there are likely going to be a ton of guards put everywhere now so who knows what that impact will be. We'll just take a different performance hit as the work needs to be fundamentally done. Dintner elsewhere will hopefully involve less tight loops.

I'll pull this part out of the PR but there is likely a very significant regression in the inability to handle malformed macro and other inputs after the removal of no nil seqs. Simply saying use safeLen which itself is buggy (should have the updated case on kind) and will need to be changed. checkMinSons and the like will more than likely have to use safeLen require the introduction of that in many more areas. The better approach is likely putting safeLen in checkSons and friends.

I do agree slowing down len sucks because it's used everywhere including tight loops and this doesn't need to be the approach or is even the right approach to begin with. I do want to ensure I've clarified the pervasive nature of the issue. My hope was that you'd see a more central area or layer to detect/guard against it. Again I'll pull it out here and hopefully a better plan of attack can be sorted out elsewhere.

Araq replied:
I don't agree but let's this dicuss it elsewhere.

saem committed:
fixes #17198, DFA failure on large case stmts

This alters the DFA control flow graph generation for case statments.
Gotos are now generated as a chained link, this ensures that evaluation
of variant branches collapses as early as possible, without hitting the
2k call limit.

saem commented:
I've updated the PR removing the len related changes, I'll update that separately. I ran tests locally and they pass, so just a matter of playing the waiting game with CI.
